src/Simulation/TargetDefinitions/Decompositions/CZFromSinglyControlled.qs

```diff
@@ -41,8 +41,95 @@ namespace Microsoft.Quantum.Canon {
+                use temps = Qubit[3];
+                within {
+                    PhaseCCX(ctls[0], ctls[1], temps[0]);
+                    PhaseCCX(ctls[2], ctls[3], temps[1]);
+                    PhaseCCX(temps[0], temps[1], temps[2]);
+                }
+                apply {
+                    CCZ(temps[2], control, target);
+                }
+            }
+            elif Length(ctls) == 5 {
+                use temps = Qubit[4];
+                within {
+                    PhaseCCX(ctls[0], ctls[1], temps[0]);
+                    PhaseCCX(ctls[2], ctls[3], temps[1]);
+                    PhaseCCX(ctls[4], control, temps[2]);
+                    PhaseCCX(temps[0], temps[1], temps[3]);
+                }
+                apply {
+                    CCZ(temps[2], temps[3], target);
+                }
+            }
+            elif Length(ctls) == 6 {
+                use temps = Qubit[5];
+                within {
+                    PhaseCCX(ctls[0], ctls[1], temps[0]);
+                    PhaseCCX(ctls[2], ctls[3], temps[1]);
+                    PhaseCCX(ctls[4], ctls[5], temps[2]);
+                    PhaseCCX(temps[0], temps[1], temps[3]);
+                    PhaseCCX(temps[2], temps[3], temps[4]);
+                }
+                apply {
+                    CCZ(temps[4], control, target);
+                }
+            }
+            elif Length(ctls) == 7 {
+                use temps = Qubit[6];
+                within {
+                    PhaseCCX(ctls[0], ctls[1], temps[0]);
+                    PhaseCCX(ctls[2], ctls[3], temps[1]);
+                    PhaseCCX(ctls[4], ctls[5], temps[2]);
+                    PhaseCCX(ctls[6], control, temps[3]);
+                    PhaseCCX(temps[0], temps[1], temps[4]);
+                    PhaseCCX(temps[2], temps[3], temps[5]);
+                }
+                apply {
+                    CCZ(temps[4], temps[5], target);
+                }
+            }
+            elif Length(ctls) == 8 {
+                use temps = Qubit[7];
+                within {
+                    PhaseCCX(ctls[0], ctls[1], temps[0]);
+                    PhaseCCX(ctls[2], ctls[3], temps[1]);
+                    PhaseCCX(ctls[4], ctls[5], temps[2]);
+                    PhaseCCX(ctls[6], ctls[7], temps[3]);
+                    PhaseCCX(temps[0], temps[1], temps[4]);
+                    PhaseCCX(temps[2], temps[3], temps[5]);
+                    PhaseCCX(temps[4], temps[5], temps[6]);
+                }
+                apply {
+                    CCZ(temps[6], control, target);
+                }
+            }
             else {
-                ApplyWithLessControlsA(Controlled CZ, (ctls, (control, target)));
+                fail "Too many control qubits specified to CZ gate.";
+
+                // Eventually, we can use recursion via callables with the below utility:
+                // ApplyWithLessControlsA(Controlled Z, (ctls, qubit));
```

Comment:
Weird idea:

```suggestion
                if DecompositionsSupportBoundedRecursion() {
                    ApplyWithLessControlsA(Controlled Z, (ctls, qubit));
                }
```

Assuming a function elsewhere of the form:

```
// Config.qs
function DecompositionsSupportBoundedRecursion() : Bool {
    return false;
}
// AdvancedConfig.qs
function DecompositionsSupportBoundedRecursion() : Bool {
    return true;
}
```

Now a target that supports the advanced case can use `<Compile Include="AdvancedConfig.qs" />` while more limited targets can use `<Compile Include="Config.qs" />`, perhaps?

Not sure if it works, sorry, but an odd thought.

Reply:
It's a neat idea we could consider for the future. For now, the motivation behind avoiding recursion is to facilitate inlining, which doesn't know about optimizing across constant branches that cannot be executed (that kind of constant folding and dead code elimination would be done after inlining).